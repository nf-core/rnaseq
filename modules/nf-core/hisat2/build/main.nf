process HISAT2_BUILD {
    tag "$fasta"
    label 'process_high'
    label 'process_high_memory'

    // WARN: Version information not provided by tool on CLI. Please update version string below when bumping container versions.
    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/hisat2:2.2.1--h1b792b2_3' :
        'biocontainers/hisat2:2.2.1--h1b792b2_3' }"

    input:
    tuple val(meta), path(fasta)
    tuple val(meta2), path(gtf)
    tuple val(meta3), path(splicesites)

    output:
    tuple val(meta), path("hisat2") , emit: index
    path "versions.yml"             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def avail_mem = 0
    if (!task.memory) {
        log.info "[HISAT2 index build] Available memory not known - defaulting to 0. Specify process memory requirements to change this."
    } else {
        log.info "[HISAT2 index build] Available memory: ${task.memory}"
        avail_mem = task.memory.toGiga()
    }

    def ss = ''
    def exon = ''
    def extract_exons = ''
    def hisat2_build_memory = params.hisat2_build_memory ? (params.hisat2_build_memory as nextflow.util.MemoryUnit).toGiga() : 0
    if (avail_mem >= hisat2_build_memory) {
        log.info "[HISAT2 index build] At least ${hisat2_build_memory} GB available, so using splice sites and exons to build HISAT2 index"
        extract_exons = gtf ? "hisat2_extract_exons.py $gtf > ${gtf.baseName}.exons.txt" : ""
        ss = splicesites ? "--ss $splicesites" : ""
        exon = gtf ? "--exon ${gtf.baseName}.exons.txt" : ""
    } else {
        log.info "[HISAT2 index build] Less than ${hisat2_build_memory} GB available, so NOT using splice sites and exons to build HISAT2 index."
        log.info "[HISAT2 index build] Use --hisat2_build_memory [small number] to skip this check."
    }
    def VERSION = '2.2.1' // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
    """
    mkdir hisat2
    $extract_exons
    hisat2-build \\
        -p $task.cpus \\
        $ss \\
        $exon \\
        $args \\
        $fasta \\
        hisat2/${fasta.baseName}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        hisat2: $VERSION
    END_VERSIONS
    """
}
