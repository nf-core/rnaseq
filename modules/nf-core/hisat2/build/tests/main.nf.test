nextflow_process {

    name "Test Process HISAT2_BUILD"
    script "../main.nf"
    process "HISAT2_BUILD"
    tag "modules"
    tag "modules_nfcore"
    tag "hisat2"
    tag "hisat2/build"
    tag "hisat2/extractsplicesites"

    test("Should run without failures") {

        setup {
            run("HISAT2_EXTRACTSPLICESITES") {
                script "../../extractsplicesites/main.nf"
                process {
                """
                input[0] = Channel.of([
                    [id:'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.gtf', checkIfExists: true)
                ])
                """
                }
            }
        }

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of([
                    [id:'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[1] = Channel.of([ [id:'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.gtf', checkIfExists: true)
                ])
                input[2] = HISAT2_EXTRACTSPLICESITES.out.txt
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
