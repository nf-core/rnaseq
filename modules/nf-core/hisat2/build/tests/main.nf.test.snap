{
    "Should run without failures": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "genome"
                        },
                        [
                            "genome.1.ht2:md5,057cfa8a22b97ee9cff4c8d342498803",
                            "genome.2.ht2:md5,47b153cd1319abc88dda532462651fcf",
                            "genome.3.ht2:md5,4ed93abba181d8dfab2e303e33114777",
                            "genome.4.ht2:md5,c25be5f8b0378abf7a58c8a880b87626",
                            "genome.5.ht2:md5,91198831aaba993acac1734138c5f173",
                            "genome.6.ht2:md5,265e1284ce85686516fae5d35540994a",
                            "genome.7.ht2:md5,9013eccd91ad614d7893c739275a394f",
                            "genome.8.ht2:md5,33cdeccccebe80329f1fdbee7f5874cb"
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,e36ef3cd73d19ccf2378c9358fe942c0"
                ],
                "index": [
                    [
                        {
                            "id": "genome"
                        },
                        [
                            "genome.1.ht2:md5,057cfa8a22b97ee9cff4c8d342498803",
                            "genome.2.ht2:md5,47b153cd1319abc88dda532462651fcf",
                            "genome.3.ht2:md5,4ed93abba181d8dfab2e303e33114777",
                            "genome.4.ht2:md5,c25be5f8b0378abf7a58c8a880b87626",
                            "genome.5.ht2:md5,91198831aaba993acac1734138c5f173",
                            "genome.6.ht2:md5,265e1284ce85686516fae5d35540994a",
                            "genome.7.ht2:md5,9013eccd91ad614d7893c739275a394f",
                            "genome.8.ht2:md5,33cdeccccebe80329f1fdbee7f5874cb"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,e36ef3cd73d19ccf2378c9358fe942c0"
                ]
            }
        ],
        "timestamp": "2023-10-16T14:42:22.381609786"
    }
}